doc/man/freesasa.1.in
.\" Man page for FressSASA
.TH freesasa 1 "2017" "@PACKAGE_VERSION@" "FreeSASA man page"
.SH NAME
FreeSASA @PACKAGE_VERSION@ - calculate Solvent Accessible Surface Areas from PDB files
.SH SYNOPSIS
.B freesasa \fIPDB\-FILE\fR ... [ \-\-\fBshrake\-rupley\fR | \-\-\fBlee\-richards\fR
    \fB\-\-probe\-radius=\fR\fINUMBER\fR
    \fB\-\-resolution=\fR\fIINTEGER\fR \fB\-\-n\-threads=\fR\fIINTEGER\fR
    \fB\-\-radius\-from\-occupancy\fR | \fB\-\-config\-file=\fR\fIFILE\fR | \fB\-\-radii=\fR\fBprotor\fR|\fBnaccess\fR
    \fB\-\-separate\-models\fR | \fB\-\-join\-models\fR
    \fB\-\-hetatm\fR \fB\-\-hydrogen\fR
    \fB\-\-separate\-chains\fR | \fB\-\-chain\-groups=\fR\fISTRING\fR ...
    \fB\-\-unknown=\fR\fBguess\fR|\fBskip\fR|\fBhalt\fR
    \fB\-\-cif
    \fB\-\-output=\fR\fIFILE\fR \fB\-\-error-file=\fR\fIFILE\fR \fB\-\-no\-warnings\fR
    \fB\-\-select=\fR\fISTRING\fR ...
    \fB\-\-format=\fR\fBlog\fR|\fBres\fR|\fBseq\fR|\fBpdb\fR|\fBrsa\fR|\fBxml\fR|\fBjson\fR ...
    \fB\-\-depth\fR=\fBstructure\fR|\fBchain\fR|\fBresidue\fR|\fBatom\fR ]
.sp
.B freesasa
[\fIoptions\fR] < \fIPDB-FILE\fR
.sp
.B freesasa
(\fB\-\-help\fR | \fB\-\-version\fR | \fB\-\-deprecated\fR)
.sp

.SH DESCRIPTION
Calculate the Solvent Accessible Surface Area (SASA) of biomolecules from PDB files using either Lee & Richards' or Shrake & Rupley's algorithms.

Report bugs to:
.UR
https://github.com/mittinatten/freesasa/issues
.UE

Home page:
.UR
http://freesasa.github.io
.UE


.SH OPTIONS

.SS Parameters
.TP
.BR  \-S ", " \-\-shrake-rupley
Use Shrake & Rupley algorithm
.TP
.BR  \-L ", " \-\-lee-richards
Use Lee & Richards algorithm [default]
.TP
.BR \-p ", " \-\-probe\-radius " " \fINUMBER\fR
Set probe radius in Angstroms [default: 1.40 Å]
.TP
.BR -n ", " \-\-resolution " " \fIINTEGER\fR
Set resolution for calculation.
  S&R: number of test points/atom [default: 100],
  L&R: slices/atom [default: 20].
.TP
.BR -t ", " \-\-n\-threads " " \fIINTEGER\fR
Number of threads to use [default: 2]

.SS Atom radii and classes (maximum one of the following)
.TP
.BR \-O ", " \-\-radius\-from\-occupancy
Read atomic radii from Occupancy in PDB input
.TP
.BR \-c ", " \-\-config\-file " " \fIFILE\fR
Clasifier to determine radii and classes. Example files in 'share/'
.TP
.BR \-\-radii " " protor|naccess
Use either ProtOr or NACCESS radii and classes [defatul: protor]

.SS Input options
.TP
.BR \-\-cif
Input is in mmCIF format
.TP
.BR \-H ", " \-\-hetatm
Include HETATM entries from input
.TP
.BR \-Y ", " \-\-hydrogen
Include hydrogen atoms, suppress warnings with -w
.TP
.BR \-m ", " \-\-join\-models
Join all MODELs in input into one structure
.TP
.BR \-C ", " \-\-separate-chains
Calculate SASA for each chain separately
.TP
.BR \-M ", " \-\-separate-models
Calculate SASA for each MODEL separately
.TP
.BR \-\-unknown " " guess|skip|halt
When unknown atom is encountered, either guess its radius/class, skip it, or halt. [default: guess]
.TP
.BR \-g ", " \-\-chain\-groups " " \fILIST\fR
Each group will be treated separately.
.IP
Examples:
  '-g A', '-g A+B', '-g A -g B', '-g AB+CD'

.SS Output options
.TP
.BR \-w ", " \-\-no\-warnings
Skip most warnings
.TP
.BR \-o ", " \-\-output " " \fIFILE\fR
Output file
.TP
.BR \-e ", " \-\-error\-file " " \fIFILE\fR
Redirect errors and warnings to file
.TP
.BR -f ", " \-\-format " " log|res|seq|pdb|rsa|cif|xml|json
Output format, can be repeated. [default: log]
.TP
.BR -d ", " \-\-depth " " structure|chain|residue|atom
Depth of JSON and XML output [default: chain]
.TP
.BR \-\-select " " \fICOMMAND\fR
Select atoms using Pymol select syntax, can be repeated.
.IP
Examples:
  \-\-select "AR, resn ala+arg", \-\-select "chain_A, chain A"
.SS Deprecated
.PP
These options have been replaced and will disappear in later versions
.TP
.BR  \-\-rsa
Equivalent to \-\-format=rsa
.TP
.BR \-B  ", " \-\-print\-as\-B\-values
Equivalent to \-\-format=pdb
.TP
.BR \-r ", " \-\-foreach\-residue\-type
Equivalent to \-\-format=res
.TP
.BR \-R ", " \-\-foreach\-residue
Equivalent to \-\-format=seq
.TP
.BR \-l ", "  \-\-no\-log
Log suppressed if other format selected. Option has no effect.


CHANGELOG.md
# Changelog

FreeSASA uses semantic versioning. Changelog added for versions 2.x

## 2.1.2

- Fix error where CLI options weren't parsed properly on ARM processors.
- Fix error in man pages.

## 2.1.1

- Fix error where compiling without multithread support still set the default number of threads to 2.
- Fix issue when stripping quotes from atom names in CIF-files, which sometimes caused memory errors.

## 2.1.0

The new features in this release were created on initiative by and in collaboration
with [Danny Diaz](https://github.com/danny305).

### Added

- Support for mmCIF input with the CLI option `--cif`.
  - The CLI was ported to C++ to allow using Gemmi for CIF import.
  - Gemmi is imported as a git submodule, see README for details.
  - A test runner was added to verify that CIF and PDB input files
    give the same result.
  - The C API does not support CIF for now (this would require conversion to C++).
- Add output option `--format=cif` that can be used when input is mmCIF.
  See documentation for an example.
- Added VdW-radii for all elements.
  - Using values from Gemmi when missing in paper by Mantina et al.
  - Correct Mg radius from 1.74 to 1.73 Å.

### Fixed

- Fix bug in JSON output where relative SASA for amino acids without sidechan
  were written as `NaN`, which is not valid JSON. These values are now
  simply skipped.
- Fix bug where elements with H or D as second letter, such as CD, were classified
  as hydrogens.
- Change some type definitions to be more C++ friendly.

## 2.0.3

This version separates the Python bindings into a separate
[module](https://github.com/freesasa/freesasa-python).
To be able to release Windows binaries of the python module,
the code has been changed to be C89 compatible where necessary,
with a few macros to allow using `restrict` and `inline` when
compiled with a C99 compiler. There are no changes to the behavior
of the CLI or C API in this release.

## 2.0.2

- Relative SASA values are now calculated using the same reference
  configurations as NACCESS. This means relative SASA values will
  differ slightly from those of previous versions.
- Python bindings compatible with Python 3.
- Man page added
- Bugs fixed:
  - CLI option `--separate-models` now outputs all models in PDB output.
  - On some platforms compilation failed when libxml and/or json-c was
    not present. Now fixed.
  - Some memory allocations were not checked for failure in S&R
    calculations. These are now properly checked, and done more
    seldomly, which should improve performance slightly.
- Residue numbers now include the iCode field for insterted residues
- Compatibility with Microsoft C Compiler:
  - No variable length arrays
  - Some macros and keywords are redefined when necessary
  - Lexer does not depend on header `unistd.h`
  - GNU getline not used, replaced by fgets()

## 2.0.1

- Add function `calcCoord()` to Python interface, wraps C function
  `freesasa_calc_coord()`.
- Expand selection syntax:
  - Allow negative residue numbers `resi \-10` (escaped with
    backslash), and expression such as `resi \-10-\-5+\-3-5`
  - Allow open-ended residue ranges `resi -10`, `resi 10-` and
    combined as `resi -10+15-20+30-`
  - Allow primes in atom names (i.e. `O5'`)
  - Allow selection names to be numbers or start with a number

## 2.0

### Changed

#### General

- Classifier configuration files now have new name field

#### C API

- Classifier interface changed.
  - Pointer is now opaque and classification done via
    `freesasa_classifier_class()` instead of
    `classifier->sasa_class()`.
  - Classifiers only classify polar/apolar/unknown. Arbitrary classes
    no longer allowed.
  - The static classifiers in `freesasa.h` have reference values for
    all residue types, allowing calculation of relative SASA for for
    example RSA output. At the moment no support for defining these
    values in classifier configuration file (TODO?).
  - Interface for classifying using string-value pairs has been
    removed (type `freesasa_strvp` and functions
    `freesasa_result_classifiy()` and `freesasa_strvp_free()`). The
    new tree-interface should be used instead.
- Structure interface changed: A structure is initialized with a
  classifier and the relevant classifier values are stored directly in
  the structure. This assures that the structure and results are
  always analyzed with consistent parameters. One effect is that the
  function `freesasa_calc_structure()` no longer takes the atomic
  radii as an explicit parameters.
- The logging functions `freesasa_log()`, `freesasa_write_result()`,
  `freesasa_write_parameters()`, `freesasa_write_rsa()`,
  `freesasa_write_pdb()`, `freesasa_per_residue()`,
  `freesasa_per_residue_type()` and `freesasa_per_chain()` have been
  removed from the interface. The same functionality can be achieved
  using `freesasa_tree_export()`.

#### CLI

- CLI output format is controlled by the option `-f` or
  `--format`. Options specifying files for specific outpt types have
  been removed, and the separate options controlling output format
  have been deprecated (see below). An option `--deprecated` can be
  used to list these. This has also made the option `-l` or `--no-log`
  obsolete.

### Added

#### General

- New output formats
  - XML using libmxl2 (optional)
  - JSON using JSON-C (optional)
  - RSA (was present already in 1.1, but interface now consolidated)
- To build FreeSASA 2.0 the libraries libxml2 and JSON-C need to be
  installed. These can be disabled by `./configure --disable-json --disable-xml` to build a dependency-free version of the library.

#### C API

- New node interface. Results in tree form can be generated using
  `freesasa_calc_tree()` or `freesasa_tree_init()`. The feature was
  added to facilitate generation of XML and JSON output, but can be
  used directly to analyze results as well. The results of a
  calculation are to now best exported using `freesasa_tree_export()`.

### Fixed


src/freesasa.c
/**
 * This source file contains everything that is in freesasa.h
 * interface and does not have a natural home in any of the other
 * source files.
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif
#include <assert.h>
#include <stdlib.h>
#include <string.h>

#include "freesasa_internal.h"

#ifdef PACKAGE_VERSION
const char *freesasa_version = PACKAGE_VERSION;
#else
const char *freesasa_version = "";
#endif

#ifdef PACKAGE_STRING
const char *freesasa_string = PACKAGE_STRING;
#else
const char *freesasa_string = "FreeSASA";
#endif

/* Allows compilation with different defaults
   depending on USE_THREADS. but still exposing the value in a header
   that doesn't depend on USE_THREADS */
#if USE_THREADS
#define DEF_NUMBER_THREADS 2
#else
#define DEF_NUMBER_THREADS 1
#endif
const int FREESASA_DEF_NUMBER_THREADS = DEF_NUMBER_THREADS;

const freesasa_parameters freesasa_default_parameters = {
    FREESASA_DEF_ALGORITHM,
    FREESASA_DEF_PROBE_RADIUS,
    FREESASA_DEF_SR_N,
    FREESASA_DEF_LR_N,
    DEF_NUMBER_THREADS};

static freesasa_result *
result_new(int n)
{
    freesasa_result *result = malloc(sizeof(freesasa_result));

    if (result == NULL) {
        mem_fail();
        return NULL;
    }

    result->sasa = malloc(sizeof(double) * n);

    if (result->sasa == NULL) {
        mem_fail();
        freesasa_result_free(result);
        return NULL;
    }

    result->n_atoms = n;

    return result;
}

void freesasa_result_free(freesasa_result *r)
{
    if (r) {
        free(r->sasa);
        free(r);
    }
}

freesasa_result *
freesasa_calc(const coord_t *c,
              const double *radii,
              const freesasa_parameters *parameters)

{
    freesasa_result *result;
    int ret = FREESASA_SUCCESS, i;

    assert(c);
    assert(radii);

    result = result_new(freesasa_coord_n(c));

    if (result == NULL) {
        fail_msg("");
        return NULL;
    }

    if (parameters == NULL) parameters = &freesasa_default_parameters;

    switch (parameters->alg) {
    case FREESASA_SHRAKE_RUPLEY:
        ret = freesasa_shrake_rupley(result->sasa, c, radii, parameters);
        break;
    case FREESASA_LEE_RICHARDS:
        ret = freesasa_lee_richards(result->sasa, c, radii, parameters);
        break;
    default:
        assert(0); /* should never get here */
        break;
    }
    if (ret == FREESASA_FAIL) {
        freesasa_result_free(result);
        return NULL;
    }

    result->total = 0;
    for (i = 0; i < freesasa_coord_n(c); ++i) {
        result->total += result->sasa[i];
    }
    result->parameters = *parameters;

    return result;
}

freesasa_result *
freesasa_calc_coord(const double *xyz,
                    const double *radii,
                    int n,
                    const freesasa_parameters *parameters)
{
    coord_t *coord = NULL;
    freesasa_result *result = NULL;

    assert(xyz);
    assert(radii);
    assert(n > 0);

    coord = freesasa_coord_new_linked(xyz, n);
    if (coord != NULL) result = freesasa_calc(coord, radii, parameters);
    if (result == NULL) fail_msg("");

    freesasa_coord_free(coord);

    return result;
}

freesasa_result *
freesasa_calc_structure(const freesasa_structure *structure,
                        const freesasa_parameters *parameters)
{
    assert(structure);

    return freesasa_calc(freesasa_structure_xyz(structure),
                         freesasa_structure_radius(structure),
                         parameters);
}

freesasa_node *
freesasa_calc_tree(const freesasa_structure *structure,
                   const freesasa_parameters *parameters,
                   const char *name)
{
    freesasa_node *tree = NULL;
    freesasa_result *result;

    assert(structure);

    result = freesasa_calc(freesasa_structure_xyz(structure),
                           freesasa_structure_radius(structure),
                           parameters);

    if (result != NULL) {
        tree = freesasa_tree_init(result, structure, name);
    } else {
        fail_msg("");
    }

    if (tree == NULL) {
        fail_msg("");
    }

    freesasa_result_free(result);

    return tree;
}

static inline void
count_err(int return_value, int *n_err)
{
    if (return_value == FREESASA_FAIL) {
        (*n_err)++;
    }
}

int freesasa_tree_export(FILE *file,
                         freesasa_node *root,
                         int options)
{
    int n_err = 0;

    assert(freesasa_node_type(root) == FREESASA_NODE_ROOT);

    if (options & FREESASA_LOG) {
        count_err(freesasa_write_log(file, root), &n_err);
    }
    if (options & FREESASA_RES) {
        count_err(freesasa_write_res(file, root), &n_err);
    }
    if (options & FREESASA_SEQ) {
        count_err(freesasa_write_seq(file, root), &n_err);
    }
    if (options & FREESASA_PDB) {
        count_err(freesasa_write_pdb(file, root), &n_err);
    }
    if (options & FREESASA_RSA) {
        count_err(freesasa_write_rsa(file, root, options), &n_err);
    }
    if (options & FREESASA_JSON) {
#if USE_JSON
        count_err(freesasa_write_json(file, root, options), &n_err);
#else
        return fail_msg("library was built without support for JSON output");
#endif
    }
    if (options & FREESASA_XML) {
#if USE_XML
        count_err(freesasa_write_xml(file, root, options), &n_err);
#else
        return fail_msg("library was built without support for XML output");
#endif
    }
    if (n_err > 0) {
        return fail_msg("there were errors when writing output");
    }
    return FREESASA_SUCCESS;
}

freesasa_result *
freesasa_result_clone(const freesasa_result *result)
{
    freesasa_result *clone = result_new(result->n_atoms);

    if (clone == NULL) {
        fail_msg("");
        return NULL;
    }

    clone->n_atoms = result->n_atoms;
    clone->total = result->total;
    clone->parameters = result->parameters;
    memcpy(clone->sasa, result->sasa, sizeof(double) * clone->n_atoms);

    return clone;
}

const char *
freesasa_alg_name(freesasa_algorithm alg)
{
    switch (alg) {
    case FREESASA_SHRAKE_RUPLEY:
        return "Shrake & Rupley";
    case FREESASA_LEE_RICHARDS:
        return "Lee & Richards";
    default:
        // This should never happen
        assert(0 && "Illegal algorithm");
        return "Unknown algorithm";
    }
}
